airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java

```diff
@@ -22,97 +24,95 @@
+    ContainerOrException(Supplier<GenericContainer<?>> containerSupplier) {
+      this.containerSupplier = containerSupplier;
+    }
 
-    @SuppressWarnings("unchecked")
-    static private <C extends JdbcDatabaseContainer<?>> C getOrCreate(String mapKey, ContainerFactory<C> factory) {
-      final Singleton<?> singleton = LAZY.computeIfAbsent(mapKey, Singleton<C>::new);
-      return ((Singleton<C>) singleton).getOrCreate(factory);
+    GenericContainer<?> container() {
+      if (exception == null && _container == null) {
+        synchronized (this) {
+          if (_container == null && exception == null) {
+            try {
+              _container = containerSupplier.get();
+            } catch (RuntimeException e) {
+              exception = e;
+            }
+          }
+        }
+      }
+      return _container;
     }
 
-    final private String imageName;
-    final private List<String> methodNames;
+  }
 
-    private C sharedContainer;
-    private RuntimeException containerCreationError;
+  private static final ConcurrentMap<ContainerKey, ContainerOrException> SHARED_CONTAINERS = new ConcurrentHashMap<>();
 
-    private Singleton(String imageNamePlusMethods) {
-      final String[] parts = imageNamePlusMethods.split("\\+");
-      this.imageName = parts[0];
-      this.methodNames = Arrays.stream(parts).skip(2).toList();
-    }
+  /**
+   * Creates a new, unshared testcontainer instance. This usually wraps the default constructor for
+   * the testcontainer type.
+   */
+  protected abstract C createNewContainer(DockerImageName imageName);
 
-    private synchronized C getOrCreate(ContainerFactory<C> factory) {
-      if (sharedContainer == null && containerCreationError == null) {
-        try {
-          create(imageName, factory, methodNames);
-        } catch (RuntimeException e) {
-          sharedContainer = null;
-          containerCreationError = e;
-        }
-      }
-      if (containerCreationError != null) {
-        throw new RuntimeException(
-            "Error during container creation for imageName=" + imageName
-                + ", factory=" + factory.getClass().getName()
-                + ", methods=" + methodNames,
-            containerCreationError);
-      }
-      return sharedContainer;
+  /**
+   * Returns a shared instance of the testcontainer.
+   */
+  public final C shared(String imageName, String... methods) {
+    List<String> methodList = methods == null ? Collections.emptyList() : Arrays.asList(methods);
+    DockerImageName dockerImageName = DockerImageName.parse(imageName);
+    final ContainerKey containerKey = new ContainerKey(getClass(), dockerImageName, methodList);
+    ContainerOrException containerOrError = SHARED_CONTAINERS.computeIfAbsent(containerKey, this::createContainerOrError);
+    if (containerOrError.container() == null) {
+      throw containerOrError.exception;
     }
```

Comment:
Wouldn't this throw be better located in the `container()` method? I mention this because having it here is a bit surprising, reading that method definition it feels like it's missing there. Perhaps it's just me. It's not very important.

Reply:
Now I see what you mean. Makes complete sense. Thank you!

Reply:
(duh)